### PR TITLE
cpio and zstd

### PR DESCRIPTION
**Pull request by keszybz**
This adds support for building cpio.zstd archives. See https://github.com/keszybz/mkosi-initrd for instructions how to create a functional pure-systemd initrd from Fedora packages.

I'm marking it as RFC. The cpio part is fine I think. The *implementation* of compression with zstd is also OK. But the user interface needs some design: right now we have `--xz` option, based on the assumption that `xz` is the only compression one would want to use. I think this isn't really true any more: `zstd` achieves very similar compression at a fraction of the compression/decompression times. For images that are widely distributed `xz` is still probably the best choice, but `zstd` is better for local work and anything where speed matters.

So I think we should deprecate the `--xz` option, and have `--inner-compression` (internally in the fs or archive where supported) and `--outer-compression` (for doing compression of the full output, like `--xz` right now), and `--compression` that would pick `--inner-compression` or `--outer-compression` as appropriate. Most of the time users would specify `--compression=yes` or `--compression=zstd|xz|gz|…` and `mkosi` would apply it in the appropriate place.

If you think that's the way to go, I will rework the pull request along those lines.

### CODE REVIEW

**Comment by behrmann** [comment]
Very cool!

We already hat a PR about adding `--zstd` in addition to `--xz`, but it stalled.

**Comment by keszybz** [comment]
That was #465. Yeah, I don't think another `--zstd` switch is the way to go. The "best" compression algorithm changes surprisingly often.

**Comment by poettering** [comment]
not a fan of the "inner" and "outer" compression. I'd be precise here, and call one "fs" compression an the other "archive" compression. Naming is hard...

File: mkosi/__init__.py
Comment (poettering) [inline mkosi/__init__.py:6085]: split out this one into a separate PR and merge right away?

File: mkosi/__init__.py
Comment (poettering) [inline mkosi/__init__.py:3729]: Maybe rename the function make_tar() → make_archive(). The log message already talks about "archive" anyway...

File: mkosi/__init__.py
Comment (poettering) [inline mkosi/__init__.py]: share the infra from xz_binary() here to use pxz only if installed, and use plain xz otherwise

File: mkosi/__init__.py
Comment (poettering) [inline mkosi/__init__.py:2309]: isn't there an official way to get rid of this in fedora?

File: mkosi/__init__.py
Comment (poettering) [inline mkosi/__init__.py:3759]: shouldn't we share this with the tar codepaths? i.e. if we invoke the compressor manually for the cpio case there shouldn't be any reason why we couldn't do this for the tar case the exact same way?

File: mkosi/__init__.py
Comment (keszybz) [reply mkosi/__init__.py]: That chunk is replaced later on with something that does the check…

File: mkosi/__init__.py
Comment (keszybz) [reply mkosi/__init__.py:2309]: Not directly. `rpm` has `--excludepaths`, but I'm not sure if there's a way to tell dnf to pass that.

File: mkosi/__init__.py
Comment (keszybz) [reply mkosi/__init__.py:3759]: Yeah, it would be nice to share this. I didn't do that so far because that'll require a bigger restructuring of how `tar` is invoked.

File: mkosi/__init__.py
Comment (keszybz) [reply mkosi/__init__.py:3729]: I changed tar→archive in various places to make the code more similar in `make_tar` and `make_cpio`. But this function should still be `make_tar`, because it creates a specific type of archive, and there are other functions for other types.

File: mkosi/__init__.py
Comment (keszybz) [reply mkosi/__init__.py:6085]: It's interleaved with subsequent changes… I had to make a few attempts to have sentence structure which works well with templating of all the options.

**Comment by keszybz** [comment]
> not a fan of the "inner" and "outer" compression. I'd be precise here, and call one "fs" compression an the other "archive" compression. Naming is hard...

So... let's say we add EROFS support. (I think this could be a nicer alternative to squashfs images.)
EROFS has inner compression (lz4 only?). It might be useful to add outer compression with `xz` or `zstd` to shave off some bytes for distribution. It doesn't fit to call the outer compression "archive" in this case, because there's not archive involved.

File: mkosi/__init__.py
Comment (keszybz) [reply mkosi/__init__.py:2309]: https://bugzilla.redhat.com/show_bug.cgi?id=1967970

**Comment by keszybz** [comment]
I think this is mostly ready. The last commit is the compression part, significantly reworked along the suggested lines.

**Comment by keszybz** [comment]
```
Run python3 -m mypy mkosi
mkosi/__init__.py:3871:21: error: Item "None" of "Optional[IO[Any]]" has no
attribute "write"  [union-attr]
                        cpio.stdin.write(file.encode("utf8") + b"\0")
                        ^
mkosi/__init__.py:3872:17: error: Item "None" of "Optional[IO[Any]]" has no
attribute "close"  [union-attr]
                    cpio.stdin.close()
                    ^
Found 2 errors in 1 file (checked 3 source files)
Error: Process completed with exit code 1.
```
I don't get this here… Seems bogus.

**Comment by behrmann** [comment]
> ```
> Run python3 -m mypy mkosi
> mkosi/__init__.py:3871:21: error: Item "None" of "Optional[IO[Any]]" has no
> attribute "write"  [union-attr]
>                         cpio.stdin.write(file.encode("utf8") + b"\0")
>                         ^
> mkosi/__init__.py:3872:17: error: Item "None" of "Optional[IO[Any]]" has no
> attribute "close"  [union-attr]
>                     cpio.stdin.close()
>                     ^
> Found 2 errors in 1 file (checked 3 source files)
> Error: Process completed with exit code 1.
> ```
> 
> I don't get this here… Seems bogus.

This is semi-bogus, I think. I'm not sure what the idea is, maybe they think, that the subprocess could have closed stdin? Adding a `assert cpio.stdin is not None` below resolves the issue.

**Comment by behrmann** [comment]
Ah, I only just saw, that this is a new `popen` function. Yes, I guess is that mypy can't properly determine the type of `stdin` in the Popen and throws this error, because it could be `None`.

**Comment by keszybz** [comment]
But I'm passing `stdin=subprocess.PIPE`, so we know that it is not None. Even if the file was closed, the object is not replaced by None, just a call to `.write()` would return an error. I will add the `assert` as suggested if that is necessary to resolve the warning, but I don't think mypy is being smart here. (I'll wait to see if there are any other comments though. I don't think it's worth to restart the CI just for that.)

**Comment by behrmann** [comment]
That's true and it's clearly a bug. Would you care to open on against [mypy](https://github.com/python/mypy/issues)?

You can also add an appropriate type ignore comment to mollify mypy.

**Comment by keszybz** [comment]
https://github.com/python/mypy/issues/10583

**Comment by keszybz** [comment]
Now with a workaround for mypy.

I was experimenting with redirecting stdin from /dev/null for subprocesses. It doesn't help with the mypy error, but I think it's good to do that anyway, so I left that as a separate commit.

**Review by behrmann — COMMENTED** [review]
Only did a quick readthrough and haven't tested this yet. I like the new option names a lot better. 

I think it would make sense to also add some test for the cpio output, so that we don't break it by chance at some future point.

**Comment by poettering** [comment]
> So... let's say we add EROFS support. (I think this could be a nicer alternative to squashfs images.)
> EROFS has inner compression (lz4 only?). It might be useful to add outer compression with `xz` or `zstd` to shave off some bytes for distribution. It doesn't fit to call the outer compression "archive" in this case, because there's not archive involved.

The gzip manpage uses the word "archive" when talking about anything *.gz.  The "xz" manpaged uses it generically too to some point, but less so. (zstd man page doesn't mention the word at all). Hence I think it would not be too far off to say that a disk image that is compressed with one of the three compressors at the end, without a tar envelope is still an archive.

File: mkosi/__init__.py
Comment (poettering) [inline mkosi/__init__.py:3730]: (btw, i think to keep things systematic and well-defined on all distros we should always use "gtar" here first, and fall back to "tar", but never use bsdtar. In particular as bsdtar has no selinux control iirc)

but that's something to fix separately.

File: man/mkosi.1
Comment (poettering) [inline man/mkosi.1:406]: "output" sounds good to me too, maybe indeed better than "archive"

**Comment by poettering** [comment]
looks all good to me, but still no the python guru who could say overly smart things about the pythonicness of the code

File: mkosi/__init__.py
Comment (poettering) [reply mkosi/__init__.py:2388]: hmm, maybe we should add a new generic RemoveFiles= stanza that works on all distros, and can be used here? It would just list paths (with wildcards) to remove as very last step of the build or so?

File: mkosi/__init__.py
Comment (keszybz) [reply mkosi/__init__.py:279]: I did it because there's a `run()` function, which invokes `subprocess.run()`. So a new function which invokes `subprocess.Popen()` is called `popen()`. But I don't think the `Popen()` name is very good, so I'm happy to rename this. Maybe `spawn` ?

Comment (behrmann) [reply mkosi/__init__.py:279]: I think I like spawn.

File: mkosi/__init__.py
Comment (poettering) [reply mkosi/__init__.py:2388]: And then maybe explicit provide an example in the docs, that it's a good idea to list /usr/share/locale in this setting

File: mkosi/__init__.py
Comment (keszybz) [reply mkosi/__init__.py:3809]: Indeed, this was a left-over from previous times.

File: mkosi/__init__.py
Comment (keszybz) [reply mkosi/__init__.py]: I rewrote the two functions a bit: now there's spawn() which does all the work, and run() which is a thin wrapper. This way it's not necessary to duplicate the logging and argument mangling logic.

**Comment by keszybz** [comment]
Updated with the requested changes.
I didn't add a test though: I'll work in that next.

File: mkosi/__init__.py
Comment (keszybz) [reply mkosi/__init__.py:2375]: `RemoveFiles=` sounds good. But let's do that separately, this one is already getting rather complex.

File: mkosi/__init__.py
Comment (keszybz) [reply mkosi/__init__.py]: I rewrote the wrapper again. I had trouble getting all the stdin/stdout/input/output/stderr/text cases right.

File: mkosi/__init__.py
Comment (behrmann) [inline mkosi/__init__.py:304]: This can potentially byte a user, since `input` and `stdin` cannot be non-`None` at the same time. It might make more sense to standardise on either in our wrapper.

File: mkosi/__init__.py
Comment (behrmann) [inline mkosi/__init__.py:3844]: Is there an advantage not to use tar's `--use-compress-program` as `f"--use-compress-program='{compressor}'"`? The short option is different between GNU tar and BSD tar, but the long option has the same name.

File: mkosi/__init__.py
Comment (keszybz) [reply mkosi/__init__.py:304]: There's a check above, and I think that should be enough. Standardizing on one of them is not easy, because they have completely different types and semantics...

File: mkosi/__init__.py
Comment (keszybz) [reply mkosi/__init__.py:3844]: Indeed. I'll rework it like that.

File: mkosi/__init__.py
Comment (behrmann) [reply mkosi/__init__.py:304]: True, the thought came from the above check being for both not being `None`, but in the end this will be caught in `subprocess.run` itself, so that's fine.

**Comment by keszybz** [comment]
Updated with two changes:
- `tar --use-compress-program=` is used instead of building a pipeline internally
- I dropped the commit to redirect stdin from `/dev/null`. I didn't investigate the details, but I think this was causing the CI failure. I'll submit that separately once dust settles.

**Comment by keszybz** [comment]
CI passes, yay. Let's give it more to chew.

**Comment by keszybz** [comment]
Rebased for the backend.py split.

**Comment by behrmann** [comment]
> Rebased for the backend.py split.

Thanks! I almost feared I would have to pick out the things in #729 again :)

I'll have a look at this PR later today.

File: mkosi/__init__.py
Comment (behrmann) [inline mkosi/__init__.py:5411]: Just a stylistic quibble, feel free to ignore it
```suggestion
    return f"{output}{suffix}.{compress or ''}"
```

**Review by behrmann — APPROVED** [review]
LGTM. Haven't run it myself, but there's little I could do that the tests don't already do. 

@DaanDeMeyer Can you have a look at this so that another pair of eyes looked this over?

File: mkosi/__init__.py
Comment (keszybz) [reply mkosi/__init__.py:5411]: No, that makes the dot unconditional, but it should be added only if the compression suffix is added.

Comment (behrmann) [reply mkosi/__init__.py:5411]: Right, chalk it up to post-vaccine grogginess :)

**Comment by keszybz** [comment]
@DaanDeMeyer any comments?

**Comment by DaanDeMeyer** [comment]
No time to give this a thorough look! But the changes look good and given @behrmann gave this a thorough look, I think we can merge.

(Given that I'm swamped with other things at the moment, feel free to merge things yourself @keszybz @poettering if I take too long)

**Pull request merged**
